core/iwasm/libraries/wasi-nn/src/wasi_nn.c

```diff
@@ -0,0 +1,217 @@
+    }
+
+    wasm_module_inst_t instance = wasm_runtime_get_module_inst(exec_env);
+    bh_assert(instance);
+
+    error res;
+    graph_builder_array builder_native;
+    if (success
+        != (res = graph_builder_array_app_native(instance, builder,
+                                                 &builder_native)))
+        return res;
+
+    if (!wasm_runtime_validate_native_addr(instance, graph, sizeof(graph))) {
+        NN_ERR_PRINTF("graph is invalid");
+        return invalid_argument;
+    }
+
+    current_encoding = encoding;
+    is_initialized = true;
+
+    res = lookup[current_encoding].load(&builder_native, current_encoding,
+                                        target, graph);
+
+    // XXX: Free intermediate structure pointers
+    wasm_runtime_free(builder_native.buf);
+
+    NN_DBG_PRINTF("wasi_nn_load finished with status %d [graph=%d]", res,
+                  *graph);
+
+    return res;
+}
+
+error
+wasi_nn_init_execution_context(wasm_exec_env_t exec_env, graph graph,
+                               graph_execution_context *ctx)
+{
+    NN_DBG_PRINTF("Running wasi_nn_init_execution_context [graph=%d]...",
+                  graph);
+    error res;
+    if (success != (res = is_model_initialized()))
+        return res;
+
+    res = lookup[current_encoding].init_execution_context(graph, ctx);
+    *ctx = graph;
+    NN_DBG_PRINTF(
+        "wasi_nn_init_execution_context finished with status %d [ctx=%d]", res,
+        *ctx);
+    return res;
+}
+
+error
+wasi_nn_set_input(wasm_exec_env_t exec_env, graph_execution_context ctx,
+                  uint32_t index, tensor_wasm *input_tensor)
+{
+    NN_DBG_PRINTF("Running wasi_nn_set_input [ctx=%d, index=%d]...", ctx,
+                  index);
+
+    error res;
+    if (success != (res = is_model_initialized()))
+        return res;
+
+    wasm_module_inst_t instance = wasm_runtime_get_module_inst(exec_env);
+    bh_assert(instance);
+
+    tensor input_tensor_native;
+    if (success
+        != (res = tensor_app_native(instance, input_tensor,
+                                    &input_tensor_native)))
+        return res;
+
+    res = lookup[current_encoding].set_input(ctx, index, &input_tensor_native);
+
+    // XXX: Free intermediate structure pointers
+    wasm_runtime_free(input_tensor_native.dimensions);
+
+    NN_DBG_PRINTF("wasi_nn_set_input finished with status %d", res);
+    return res;
+}
+
+error
+wasi_nn_compute(wasm_exec_env_t exec_env, graph_execution_context ctx)
+{
+    NN_DBG_PRINTF("Running wasi_nn_compute [ctx=%d]...", ctx);
+    error res;
+    if (success != (res = is_model_initialized()))
+        return res;
+
+    res = lookup[current_encoding].compute(ctx);
+    NN_DBG_PRINTF("wasi_nn_compute finished with status %d", res);
+    return res;
+}
+
+error
+wasi_nn_get_output(wasm_exec_env_t exec_env, graph_execution_context ctx,
+                   uint32_t index, tensor_data output_tensor,
+                   uint32_t *output_tensor_size)
```

Comment:
Had better use `wasm_runtime_validate_native_addr(instance, output_tensor_size, sizeof(uint32_t))` to check output_tensor_size, since for pointer argument, runtime only checks one byte of it before calling the native API.

Reply:
You are totally right, I'm fixing it...